Prevent endpoint canceling AM messages multiple times

pentschev opened the pull request:
This is important because UCX may reuse the same endpoint handle, and if a message is canceled during then endpoint finalizer, a message received on the same (new) endpoint handle may be canceled incorrectly.

Due to this, `test_close_after_n_recv[am]` would fail non-deterministically.

madsbk commented on ucp/_libs/ucx_endpoint.pyx:
```diff
@@ -54,6 +54,13 @@ cdef int _cancel_am_recv(UCXWorker worker, uintptr_t handle_as_int):
                 **cb_kwargs
             )
 
+        # Prevent endpoint canceling AM messages multiple times.  This is important
+        # because UCX may reuse the same endpoint handle, and if a message is canceled
+        # during then endpoint finalizer, a message received on the same (new) endpoint
+        # handle may be canceled incorrectly.
+        del worker._am_recv_wait[handle_as_int]
+        worker._inflight_msgs_to_cancel["am"].discard(handle_as_int)
```
```suggestion
        # Prevent endpoint canceling AM messages multiple times. This is important
        # because UCX may reuse the same endpoint handle, and if a message is canceled
        # during then endpoint finalizer, a message received on the same (new) endpoint
        # handle may be canceled incorrectly.
        del worker._am_recv_wait[handle_as_int]
        worker._inflight_msgs_to_cancel["am"].discard(handle_as_int)
```

pentschev replied:
Ha! Good catch!

madsbk reviewed (approved):
LGTM, only a very minor suggestion :P

pentschev commented:
I just realized I'm changing the size of `self._inflight_msgs_to_cancel["am"]` set during a loop with this change. I'll need to rework it a little to prevent such warnings and will push a fix to this shortly.

pentschev commented:
@madsbk I now fixed the issue I mentioned above, could you take another look if you have the chance?

madsbk commented:
> @madsbk I now fixed the issue I mentioned above, could you take another look if you have the chance?

Looks good but it is possible to write a  regression test of the mentioned issue?

pentschev commented:
> Looks good but it is possible to write a regression test of the mentioned issue?

I've added a couple of tests now, please let me know if you think they're satisfactory. It's difficult to test the exact behavior of `UCXWorker.cancel_inflight_messages` from the async API because of the lack of control to progress the worker, but I added a test for that in the core library.

madsbk commented on ucp/_libs/ucx_endpoint.pyx:
```diff
@@ -54,6 +65,27 @@ cdef int _cancel_am_recv(UCXWorker worker, uintptr_t handle_as_int):
+
+            # Prevent endpoint canceling AM messages multiple times. This is important
+            # because UCX may reuse the same endpoint handle, and if a message is
+            # canceled # during then endpoint finalizer, a message received on the same
```
```suggestion
            # canceled during then endpoint finalizer, a message received on the same
```

madsbk commented on tests/test_endpoint.py:
```diff
@@ -34,3 +34,32 @@ async def client_node(port):
+                await ep.recv(msg)
+        await ep.close()
+
+    listener = ucp.create_listener(server_node,)
```
```suggestion
    listener = ucp.create_listener(server_node)
```

madsbk commented on tests/test_endpoint.py:
```diff
@@ -34,3 +34,32 @@ async def client_node(port):
+        await ep.close()
+
+    async def client_node(port):
+        ep = await ucp.create_endpoint(ucp.get_address(), port,)
```
```suggestion
        ep = await ucp.create_endpoint(ucp.get_address(), port)
```

madsbk commented on ucp/_libs/tests/test_cancel.py:
```diff
@@ -0,0 +1,108 @@
+        )
+
+    listener = ucx_api.UCXListener(worker=worker, port=0, cb_func=_listener_handler)
+    queue.put(listener.port),
```
```suggestion
    queue.put(listener.port)
```

madsbk reviewed (approved):
> I've added a couple of tests now, please let me know if you think they're satisfactory. It's difficult to test the exact behavior of UCXWorker.cancel_inflight_messages from the async API because of the lack of control to progress the worker, but I added a test for that in the core library.

Looks great @pentschev, only have some cosmetic suggestions.

pentschev commented:
Nice catches @madsbk , I'm actually surprised `black` didn't fix those, I often just ignore them in the expectation `black` will fix them, but it doesn't seem that it does. Thanks for the review and approval!
